src/search/algorithms.clj

```diff
@@ -63,6 +63,35 @@
             new-visited (clojure.set/union children visited)]
         (recur new-frontier new-came-from new-visited)))))
 
+(defn a-star-search [children-fn cost-fn heuristic-fn start-state goal-state & {:keys [max-states] :or {max-states 1000000}}]
+  (loop [max-states max-states
+        frontier (pm/priority-map start-state 0)
+         came-from {}
+         cost-so-far {start-state 0}]
+    (if (or (neg? max-states)
+            (empty? frontier)
+            (>= (count came-from) max-states)
+            (= (first (peek frontier)) goal-state))
+      [came-from cost-so-far]
+      (let [current (first (peek frontier))
+            current-cost (cost-so-far current)
+            children (set (children-fn current))
+            children-costs (reduce #(assoc %1 %2 (+ current-cost (cost-fn current %2))) {} children)
+            children-to-add (filter #(or (not (contains? cost-so-far %))
+                                         (< (children-costs %) (cost-so-far %))) children)
+            heuristics (map (partial heuristic-fn goal-state) children-to-add)
+            new-cost-so-far (reduce #(assoc %1 %2 (children-costs %2)) cost-so-far children-to-add)
+
+            ;heuristic multiplier of 215 because everything at 215 and above returns the lowest number of states
+            new-frontier (reduce #(assoc %1 %2 (+ (* (heuristic-fn goal-state %2) 215) (children-costs %2))) (pop frontier) children-to-add)
```

Comment:
This 215 is a pretty crazy magic number. 😜  I understand how you came across it, but I would have probably `def`ed it so it had a name and you had a better place to document what was going on there.